### PR TITLE
Unify PoW implementation context

### PR DESCRIPTION
**Pull request by furuame**
`implcontext.h` provides a set of API for PoW Implementations for the new interface #57 :
* registerImplContext(ImplContext *impl_ctx);
    * Register the PoW implementation into the linked list and the `dcurl_entry` would iterate the linked list to find the available implementation.
* initializeImplContext(ImplContext *impl_ctx);
    * Initialize the context of PoW implmentation, which is provided by PoW implementation.
* destroyImplContext(ImplContext *impl_ctx);
    * Free the allocated memory by PoW implementation,  which is also provided by PoW implementation.
* enterImplContext(ImplContext *impl_ctx);
    * Multi-thread management, which enables the thread get the PoW context
* exitImplContext(ImplContext *impl_ctx);
    * Multi-thread management, which exits the critical section and let the other thread in
* getPoWContext(ImplContext *impl_ctx, int8_t *trytes, int mwm);
    * Get the PoW context, which is provided by PoW implementation
* doThePoW(ImplContext *impl_ctx, void *pow_ctx);
    * Do the PoW
* freePoWContext(ImplContext *impl_ctx, void *pow_ctx);
    * Free the PoW context
* getPoWResult(ImplContext *impl_ctx, void *pow_ctx);
    * Get the calculated trytes from PoW context

In this pull request, pure C, SSE, AVX and OpenCL implementations are updated for this new interface and pass the tests.
```
cwei@creeper ~/w/dcurl> make BUILD_GPU=1 BUILD_DEBUG=1 check
*** Validating build/test-trinary ***
	[ Verified ]
*** Validating build/test-curl ***
	[ Verified ]
*** Validating build/test-dcurl ***
	[ Verified ]
*** Validating build/test-multi-pow ***
	[ Verified ]
*** Validating build/test-pow_sse ***
	[ Verified ]
*** Validating build/test-pow_cl ***
	[ Verified ]
```

The new interface also enables multi-gpu support #24.
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.77                 Driver Version: 390.77                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1070    Off  | 00000000:01:00.0 Off |                  N/A |
| 11%   57C    P2   103W / 151W |    445MiB /  8117MiB |     95%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 1070    Off  | 00000000:02:00.0 Off |                  N/A |
|  8%   55C    P2    79W / 151W |    445MiB /  8119MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     22720      C   /usr/bin/python3                             435MiB |
|    1     22720      C   /usr/bin/python3                             435MiB |
+-----------------------------------------------------------------------------+
```

### CODE REVIEW

**Comment by jserv** [comment]
Can you rebase?

**Comment by furuame** [comment]
done.

**Comment by chenwei-tw** [comment]
@jserv Is there any review needed?

File: src/clcontext.c [hunk 1]
```diff
@@ -9,26 +9,14 @@
 #include <stdio.h>
 #include "pearl.cl.h"
 
+#define HASH_LENGTH 243               // trits
```
Comment (jserv) [inline src/clcontext.c]: Why don't you refer these macro in `src/constants.h`?

File: src/clcontext.c [hunk 2]
```diff
@@ -132,16 +120,73 @@ int init_cl_buffer(CLContext *ctx)
     return 1;
 }
 
-int init_clcontext(CLContext **ctx)
+static int init_BufferInfo(CLContext *ctx)
 {
-    *ctx = (CLContext *) malloc(sizeof(CLContext));
+    ctx->kernel_info.buffer_info[0] =
```
Comment (jserv) [inline src/clcontext.c]: These indexes look a bit ugly. Can you rewrite by mnemonic enum?

File: src/clcontext.c [hunk 3]
```diff
@@ -132,16 +120,73 @@ int init_cl_buffer(CLContext *ctx)
...
+            ret &= set_clcontext(&ctx[ctx_idx], devices[j]);
+            ret &= init_cl_kernel(&ctx[ctx_idx]);
+            ret &= init_BufferInfo(&ctx[ctx_idx]);
+            if (!ret) goto exit;
+            ctx_idx++;
+        }
+        free(devices);
+    }
+    free(platform);
+exit:
```
Comment (jserv) [inline src/clcontext.c]: Don't use `exit`, which conflicts with system call.

File: src/dcurl.c [hunk 4]
```diff
@@ -19,204 +21,97 @@
...
 #ifdef __APPLE__
 #include <dispatch/dispatch.h>
 #else
 #include <semaphore.h>
 #endif
 
-/* number of task that CPU can execute concurrently */
-static int MAX_CPU_THREAD = -1;
+/* check whether dcurl is initialized */
+static int isInitialized = 0;
```
Comment (jserv) [inline src/dcurl.c]: Use `<stdbool.h>` instead of numbers.

File: src/pow_sse.h [hunk 5]
```diff
@@ -11,13 +13,30 @@ struct _pwork_struct {
...
+    pthread_mutex_t lock;
+    pthread_t *threads;
+    Pwork_struct *pitem;
+    int8_t **nonce_array;
+    int stopSignal;
+    int num_threads;
+    /* Management of Multi-thread */
+    int indexOfContext;
+    /* Arguments of PoW */
+    int8_t input_trytes[2673]; /* 2673 */
```
Comment (jserv) [inline src/pow_sse.h]: Can we avoid the exposure of constant `2673` here?

File: src/pow_sse.c [hunk 6]
```diff
@@ -169,13 +166,13 @@ static int64_t loop128(__m128i *lmid,
                        __m128i *hmid,
                        int m,
                        int8_t *nonce,
-                       int id)
+                       int *stopSignal)
```
Comment (jserv) [inline src/pow_sse.c]: I am not sure if `stopSignal` is an effective name.

File: src/pow_cl.h [hunk 7]
```diff
@@ -2,9 +2,22 @@
...
-int8_t *PowCL(int8_t *trytes, int mwm, int index);
-int pwork_ctx_init(int context_size);
+typedef struct _pow_cl_context PoW_CL_Context;
+
+struct _pow_cl_context {
+    CLContext *clctx;
+    /* Management of Multi-thread */
+    int indexOfContext;
+    /* Arguments of PoW */
+    int8_t input_trytes[2673]; /* 2673 */
```
Comment (jserv) [inline src/pow_cl.h]: Ditto. All constants should be referred in unified header.

File: src/pow_sse.c (hunk 6 above)
Comment (furuame) [reply src/pow_sse.c]: It is used to signal other threads stop.

Comment (jserv) [reply src/pow_sse.c]: It might be ambiguous to mix `stop` and `signal` especially for people thinking it as `SIGSTOP`.

Comment (furuame) [reply src/pow_sse.c]: Okay, I will fix it.

**Pull request merged**